src/raftstore/store/snap/reader.rs

```diff
@@ -0,0 +1,307 @@
+use util::file::{calc_crc32, delete_file_if_exist, get_file_size};
+use util::rocksdb::{get_cf_handle, prepare_sst_for_ingestion, validate_sst_for_ingestion};
+
+struct SnapshotBase {
+    dir: String,
+    for_send: bool,
+    key: SnapKey,
+    snapshot_meta: SnapshotMeta,
+}
+
+impl SnapshotBase {
+    fn new(dir: String, for_send: bool, key: SnapKey, snapshot_meta: SnapshotMeta) -> Self {
+        SnapshotBase {
+            dir,
+            for_send,
+            key,
+            snapshot_meta,
+        }
+    }
+}
+
+pub struct SnapshotSender {
+    base: SnapshotBase,
+    snap_stale_notifier: Arc<SnapStaleNotifier>,
+    ref_count: Arc<AtomicUsize>,
+    sent_times: Arc<AtomicUsize>,
+
+    cur_cf_file: Option<File>,
+    cur_cf_pos: usize,
+}
+
+impl SnapshotSender {
+    pub(super) fn new(
+        dir: String,
+        key: SnapKey,
+        snapshot_meta: SnapshotMeta,
+        snap_stale_notifier: Arc<SnapStaleNotifier>,
+        ref_count: Arc<AtomicUsize>,
+        sent_times: Arc<AtomicUsize>,
+    ) -> Self {
+        let base = SnapshotBase::new(dir, true, key, snapshot_meta);
+        ref_count.fetch_add(1, Ordering::SeqCst);
+        SnapshotSender {
+            base,
+            snap_stale_notifier,
+            ref_count,
+            sent_times,
+            cur_cf_file: None,
+            cur_cf_pos: 0,
+        }
+    }
+
+    pub fn total_size(&self) -> u64 {
+        get_size_from_snapshot_meta(&self.base.snapshot_meta)
+    }
+}
+
+impl Drop for SnapshotSender {
+    fn drop(&mut self) {
+        self.ref_count.fetch_sub(1, Ordering::SeqCst);
+        self.sent_times.fetch_add(1, Ordering::SeqCst);
+    }
+}
+
+impl Read for SnapshotSender {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+        let base = &mut self.base;
+        while self.cur_cf_pos < base.snapshot_meta.get_cf_files().len() {
+            if stale_for_generate(base.key, self.snap_stale_notifier.as_ref()) {
+                let err = io::Error::new(io::ErrorKind::Other, "stale snapshot".to_owned());
+                return Err(err);
+            }
+
+            let cf = &base.snapshot_meta.get_cf_files()[self.cur_cf_pos];
+            if cf.get_size() > 0 {
+                if self.cur_cf_file.is_none() {
+                    let (dir, for_send, key) = (&base.dir, base.for_send, base.key);
+                    let cf_file_path = gen_cf_file_path(dir, for_send, key, cf.get_cf());
+                    self.cur_cf_file = Some(File::open(&cf_file_path)?);
+                }
+                match self.cur_cf_file.as_mut().unwrap().read(buf) {
+                    Ok(0) => {}
+                    Ok(n) => return Ok(n),
+                    e => return e,
+                }
+            }
+            self.cur_cf_pos += 1;
+            self.cur_cf_file = None;
+        }
+        Ok(0)
+    }
+}
+
+pub(super) struct SnapshotApplyer {
```

Comment:
is it suitable in the reader.rs?

Reply:
Both `SnapshotApplyer` and `SnapshotSender` use the same `SnapshotBase` to access disk files, so I put them in one file. Any sugesstions for better file name?